src/main/java/rx/internal/operators/OperatorBufferPredicateBoundary.java

```diff
@@ -0,0 +1,580 @@
+            return parent;
+        }
+        final BoundedBeforeSubscriber<T> parent = new BoundedBeforeSubscriber<T>(child, capacityHint, predicate, prefetch);
+        child.add(parent);
+        child.setProducer(new Producer() {
+            @Override
+            public void request(long n) {
+                parent.requestMore(n);
+            }
+        });
+        return parent;
+    }
+    
+    static final class UnboundedAfterSubscriber<T> extends Subscriber<T> {
+        final Subscriber<? super List<T>> actual;
+        
+        final int capacityHint;
+        
+        final Func1<? super T, Boolean> predicate;
+        
+        List<T> buffer;
+        
+        public UnboundedAfterSubscriber(Subscriber<? super List<T>> actual, int capacityHint,
+                Func1<? super T, Boolean> predicate) {
+            this.actual = actual;
+            this.capacityHint = capacityHint;
+            this.predicate = predicate;
+            this.buffer = new ArrayList<T>(capacityHint);
+        }
+
+        @Override
+        public void onNext(T t) {
+            List<T> b = buffer;
+            b.add(t);
+
+            boolean emit;
+            
+            try {
+                emit = predicate.call(t);
+            } catch (Throwable ex) {
+                unsubscribe();
+                Exceptions.throwOrReport(ex, this, t);
+                return;
+            }
+            
+            if (emit) {
+                actual.onNext(b);
+                buffer = new ArrayList<T>(capacityHint);
+            }
+        }
+        
+        @Override
+        public void onError(Throwable e) {
+            buffer = null;
+            actual.onError(e);
+        }
+        
+        @Override
+        public void onCompleted() {
+            List<T> b = buffer;
+            buffer = null;
+            if (!b.isEmpty()) {
+                actual.onNext(b);
+            }
+            actual.onCompleted();
+        }
+    }
+
+    static final class UnboundedBeforeSubscriber<T> extends Subscriber<T> {
+        final Subscriber<? super List<T>> actual;
+        
+        final int capacityHint;
+        
+        final Func1<? super T, Boolean> predicate;
+        
+        List<T> buffer;
+        
+        public UnboundedBeforeSubscriber(Subscriber<? super List<T>> actual, int capacityHint,
+                Func1<? super T, Boolean> predicate) {
+            this.actual = actual;
+            this.capacityHint = capacityHint;
+            this.predicate = predicate;
+            this.buffer = new ArrayList<T>(capacityHint);
+        }
+
+        @Override
+        public void onNext(T t) {
+            List<T> b = buffer;
+
+            boolean emit;
+            
+            try {
+                emit = predicate.call(t);
+            } catch (Throwable ex) {
+                unsubscribe();
+                Exceptions.throwOrReport(ex, this, t);
+                return;
+            }
+            
+            if (emit) {
```

Comment:
I think it's less confusing to not emit the buffer when it's null.
I'd add a check for `! b.isEmpty()` in the if.